utils/cache.js

```diff
@@ -0,0 +1,113 @@
+const CACHEEXPIRYTIME = 2; // Default cache expriy time in Minutes
+const CACHESIZE = 10; // Default maximum size for cache pool (in Megabytes)
+const minutesToMilliseconds = (minutes) => minutes * 60000; // Converts given minutes into milliseconds
+
+/**
+ * Cache pool to get and store API responses
+ * @param {Object} [opt] options for cache pool
+ * @param {number} opt.maximumSize Maximum size of the cache pool in Megabytes (MB)
+ */
+const cachePool = (opt = { maximumSize: CACHESIZE }) => {
+  const cacheStore = new Map();
+  let hits = 0;
+
+  /**
+   * Get an API Respnose from cacheStore.
+   * @param {string} key
+   * @returns {null | object}
+   */
+  const get = (key) => {
+    const apiData = cacheStore.get(key);
+
+    if (!apiData) {
+      return null;
+    }
+
+    const isApiDataExpired = new Date().getTime() > apiData.expiry;
+
+    // If data is expired remove it from store, time to get a fresh copy.
+    if (isApiDataExpired) {
+      cacheStore.delete(key);
+      return null;
+    }
+
+    hits += 1;
+    return apiData.response;
+  };
+
+  /**
+   * Add API response to cacheStore.
+   * @param {string} key
+   * @param {Object} value Value to be stored inside the cache.
+   * @param {number} value.priority Priority of the api
+   * @param {number} value.expiry Expiry time of api
+   * @param {number} value.size Size of api response in byte
+   * @returns {number} : statusCode
+   */
+  const set = async (key, value) => {
+    try {
+      cacheStore.set(key, value);
+      return 200;
+    } catch (err) {
+      return 500;
+    }
+  };
+
+  return { get, set, hits, cacheStore };
+};
+
+// Initialize cache pool.
+const pool = cachePool();
+
+/**
+ * Caching middleware for API resposnes.
+ * @param {Object} [data] Options to handle individual api cache.
+ * @param {number} data.priority The priority of api in cache store.
+ * @param {number} data.expiry Cache expiry time of api in minutes.
+ * @returns {function} middleware function to help cache api response.
+ */
+const cache = (data = { priority: 2, expiry: CACHEEXPIRYTIME }) => {
+  return async (req, res, next) => {
+    const key = "__cache__" + req.method + req.originalUrl;
+    const dataInCache = pool.get(key);
+
+    if (dataInCache) {
+      res.send(dataInCache);
+    } else {
+      /**
+       * As we do not have data in our cache we call the next middleware,
+       * intercept the response being sent from middleware and store it in cache.
+       *  */
+      const chunks = [];
+      const oldWrite = res.write;
+      const oldEnd = res.end;
+
+      res.write = (chunk, ...args) => {
+        chunks.push(chunk);
+        return oldWrite.apply(res, [chunk, ...args]);
+      };
+
+      res.end = (chunk, ...args) => {
+        if (chunk) {
+          chunks.push(chunk);
+        }
+
+        const apiResponse = Buffer.concat(chunks).toString();
+
+        const cacheValue = {
+          priority: data.priority,
+          response: apiResponse,
```

Comment:
The API response here seems to be in the var `apiResponse` but we are sending the whole `cacheValue` as the response on line 75, if data is found in cache? 🤔

Reply:
`pool.get` in line 72 will return the `apiResponse` and not the entire `cacheValue` so only the original response is being sent in case of a cache hit.

Reply:
Would it be better to have the `get` function return us the whole `cacheValue` and destructure it at line 72 for consistency?

Reply:
I think the current implementation of just returning the `response` is fine 👍